docs/rfcs/0004-ecs-deployment.md

```diff
@@ -0,0 +1,107 @@
+
+`ECS_CANARY_ROLLOUT` deploys workloads of the new version. it uses API such as [`DescribeServices`](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_DescribeServices.html), [`CreateTaskSet`](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_CreateTaskSet.html), [`UpdateServicePrimaryTaskSet`](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_UpdateServicePrimaryTaskSet.html) and [`DeleteTaskSet`](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_DeleteTaskSet.html).
+
+`ECS_TRAFFIC_ROUTING` changes the configuration of the load balancer you specified in .pipe.yaml in order to change the traffic routing state. It uses API such as [`DescribeTargetGroups`](https://docs.aws.amazon.com/elasticloadbalancing/latest/APIReference/API_DescribeTargetGroups.html), [`DescribeListeners`](https://docs.aws.amazon.com/elasticloadbalancing/latest/APIReference/API_DescribeListeners.html), [`ModifyListener`](https://docs.aws.amazon.com/elasticloadbalancing/latest/APIReference/API_ModifyListener.html), [`DescribeRules`](https://docs.aws.amazon.com/elasticloadbalancing/latest/APIReference/API_DescribeRules.html) and [`ModifyRule`](https://docs.aws.amazon.com/elasticloadbalancing/latest/APIReference/API_ModifyRule.html).
+
+In case of using pipeline, you need to use `External` as a deployment controller in your `serviceDefinition`.
+
+```yaml
+apiVersion: pipecd.dev/v1beta1
+kind: ECSApp
+spec:
+  input:
+    name: Sample
+    serviceDefinition: path/to/servicedef.json # optional
+    taskDefinition: path/to/taskdef.json       # required
+    loadBalancerInfo:                          # optional
+        containerName: sample-app
+        containerPort: 80
+```
+
+## Example
+
+### Canary Deployment
+
+```yaml
+apiVersion: pipecd.dev/v1beta1
+kind: ECSApp
+spec:
+  input:
+    name: Sample
+    serviceDefinition: path/to/servicedef.json
+    taskDefinition: path/to/taskdef.json
+    loadBalancerInfo:
+        containerName: sample-app
+        containerPort: 80
+  pipeline:
+    stages:
+      # Deploy workloads of the new version.
+      # But this is still receiving no traffic.
+      - name: ECS_CANARY_ROLLOUT
+      # Change the traffic routing state where
+      # the new version will receive the specified percentage of traffic.
+      # This is known as multi-phase canary strategy.
+      - name: ECS_TRAFFIC_ROUTING
+        with:
+          canary: 10
+      # Optional: We can also add an ANALYSIS stage to verify the new version.
+      # If this stage finds any not good metrics of the new version,
+      # a rollback process to the previous version will be executed.
+      - name: ANALYSIS
+      # Change the traffic routing state where
+      # thre new version will receive 100% of the traffic.
+      - name: ECS_TRAFFIC_ROUTING
+        with:
+          canary: 100
+```
```

Comment:
In this case, when the original task set (primary task set) will be terminated?

Reply:
I thought primary task set will be terminated when canary becomes 100%.

Reply:
So in the case of BlueGreen, it will also be terminated after receiving 100% traffic, right?
What if the ANALYSIS stage returns a bad result and we decide to route all traffic back to the original task set (rollback), but it was terminated previously?
Do you think that the terminating process should be done in a separate stage?

Reply:
You mean we need the stage like `K8S_CANARY_CLEAN`?
I think the stage makes pipeline more simple and easier to understand.